File: src/main/java/seremis/geninfusion/soul/entity/ai/EntityAIMoveTowardsRestrictionCustom.java
package seremis.geninfusion.soul.entity.ai;

import net.minecraft.entity.EntityLiving;
import net.minecraft.entity.ai.EntityAIBase;
import net.minecraft.util.ChunkCoordinates;
import net.minecraft.util.Vec3;
import seremis.geninfusion.api.soul.IEntitySoulCustom;
import seremis.geninfusion.api.soul.util.GIRandomPositionGenerator;

public class EntityAIMoveTowardsRestrictionCustom extends EntityAIBase {

    public IEntitySoulCustom entity;
    public EntityLiving living;

    private Vec3 target;
    private double movementSpeed;

    public EntityAIMoveTowardsRestrictionCustom(IEntitySoulCustom entity, double moveSpeed) {
        this.entity = entity;
        this.movementSpeed = moveSpeed;
        this.setMutexBits(1);
    }

    public boolean shouldExecute() {
        if(!entity.isWithinHomeDistanceCurrentPosition()) {
            ChunkCoordinates chunkcoordinates = this.entity.getHomePosition();
            target = GIRandomPositionGenerator.findRandomTargetBlockTowards(entity, 16, 7, Vec3.createVectorHelper((double) chunkcoordinates.posX, (double) chunkcoordinates.posY, (double) chunkcoordinates.posZ));

            if(target != null) {
                return true;
            }
        }
        return false;
    }

    public boolean continueExecuting() {
        return !this.living.getNavigator().noPath();
    }

    public void startExecuting() {
        this.living.getNavigator().tryMoveToXYZ(target.xCoord, target.yCoord, target.zCoord, this.movementSpeed);
    }

}

File: src/main/scala/seremis/geninfusion/soul/entity/ai/EntityAIMoveTowardsRestrictionCustom.scala
package seremis.geninfusion.soul.entity.ai

import net.minecraft.entity.EntityLiving
import net.minecraft.entity.ai.EntityAIBase
import net.minecraft.util.Vec3
import seremis.geninfusion.api.soul.IEntitySoulCustom
import seremis.geninfusion.api.util.GIRandomPositionGenerator

class EntityAIMoveTowardsRestrictionCustom(var entity: IEntitySoulCustom, moveSpeed: Double) extends EntityAIBase {

    var living: EntityLiving = entity.asInstanceOf[EntityLiving]

    private var target: Vec3 = null
    private val movementSpeed: Double = moveSpeed

    this.setMutexBits(1)

    override def shouldExecute(): Boolean = {
        if (!entity.isWithinHomeDistanceCurrentPosition_I) {
            val chunkcoordinates = this.entity.getHomePosition_I
            target = GIRandomPositionGenerator.findRandomTargetBlockTowards(entity, 16, 7, Vec3.createVectorHelper(chunkcoordinates.posX.toDouble, chunkcoordinates.posY.toDouble, chunkcoordinates.posZ.toDouble))
            if (target != null) {
                return true
            }
        }
        false
    }

    override def continueExecuting(): Boolean = !this.living.getNavigator.noPath()

    override def startExecuting() {
        this.living.getNavigator.tryMoveToXYZ(target.xCoord, target.yCoord, target.zCoord, this.movementSpeed)
    }
}